June 2022 edits from feedback

steffilazerte opened the pull request:
- [X] **Important** - Add option to subset total/composition PI results
- [X] Total - Univariate -  Give plot points tooltips (ggiraph after converted to ggplot2)
  - [x] Mention in Documentation
- [ ] Total - PI
  - [x] Decimals
  - [ ] Separate figures (once converted to ggplot2?)
- [x]  Total - Explore PI
  - [x] Fix map tooltip
  - [x] **Later/Maybe** New figure

steffilazerte committed:
Add option to subset total/composition PI results
- Add MMU_ID to meta data variables
- Create var_subset (meta data to subset by)
- Add/rearrange UIs
- Add validate catches when no groups
- Amend documentation

steffilazerte commented on inst/shiny/global.R:
```diff
@@ -30,7 +30,7 @@ var_meta <- c("SURVEY_NAM", "YT_REGION", "SURVEY_YEA", "SURVEY_ID",
               "In1Out0", "SU_STRATUM", "ALL_STRATA",
               "IDLATDEG", "IDLATMIN", "IDLONDEG", "IDLONMIN", "CENTRLAT", "CENTRLON",
-              "REGION", "GMU", "GMU2", "USE_SCALE", "SRC_SCALE", "SRC_NOTES",
+              "REGION", "MMU_ID", "GMU", "GMU2", "USE_SCALE", "SRC_SCALE", "SRC_NOTES",
```
I've added MMU_ID to the list of meta data variables. This means that it is no long available to add as a predictor, which I think is correct...?

psolymos replied:
Yes, that is correct AFAIK.

steffilazerte commented on inst/shiny/global.R:
```diff
@@ -45,6 +45,13 @@ var_filter <- c("Survey Name" = "SURVEY_NAM",
                 "Survey ID" = "SURVEY_ID")
 
+# Columns which can be subsetted (during PI exploration/summaries)
+var_subset <- c("MMU_ID", "SURVEY_NAM", "YT_REGION", "SURVEY_YEA", "SURVEY_ID",
```
These are the meta data variables available to use as subset columns in the PI subsetting... are there any here that should not be? Are there any which are missing?

psolymos replied:
Looks like there is also a `"S_SET_ID"` variant. Making it as a global variable is handy b/c then it is easy to update based on feedback.

steffilazerte committed:
Total PI moose density to 3 digits

steffilazerte committed:
Add dynamic tooltips to map
- Consider type of cell (observed/not)
- Consider type of data (model/predicted)
- Use sensible outputs depending on combination

steffilazerte commented:
@psolymos I've got the subsetting, decimal places and map tooltips up and running. You'll see that I've commented on the code in a couple of places. If you could take a look at those to make sure all is as it should be, that'd be great!

The map tooltips are a bit complex, but I wanted to make sure that we had a good combination of information. The tooltips adjust depending on what kind of variable is being mapped (observed/model data, vs. predicted data) and what cell is being looked at (a surveyed cell vs. unsurveyed cell), and whether it is a 'special variable' or not (i.e. Observed values or Cell Accuracy). Have a look and let me know what you think. 

I think I just need the ggplot figures for the rest, so I'll wait until I hear back from you. (I'm only 3.5 hrs in, so well on schedule!)

psolymos reviewed (commented):
Looks good so far! Thanks.

psolymos commented:
@steffilazerte here is the ggplot2 version of `mc_plot_univariate()`. Can you add it as an unexported function to `R/explore.R` for now?

```R
.plot_univariate <- function(i, x, dist="ZINB",
type = c("density", "map", "fit")) {

    type <- match.arg(type)

    opts <- getOption("moose_options")
    srv <- x$srv
    z <- x[,i]
    CAT <- is.factor(z) || is.character(z)
    if (CAT) {
        z <- droplevels(as.factor(z))
    }
    xy <- x[, opts$xy]
    dat <- data.frame(y=x[srv,opts$Ntot], z=z[srv])
    dat <- dat[order(dat$z),]
    m <- zeroinfl2(y ~ z | 1, dat, dist=dist, link='logit')
    dat$zhat <- stats::fitted(m)

    if (type == "density") {
      if (CAT) {
        d <- table(z)/length(srv)
        d <- data.frame(Surveyed = "All",
                        Value = names(d), n=as.numeric(d))
        d_srv <- table(z[srv])/sum(srv)
        d_srv <- data.frame(Surveyed = "Surveyed",
                            Value = names(d_srv), n=as.numeric(d_srv))
        dd <- rbind(d, d_srv)
        p <- ggplot(dd,
          aes(
            x = Value,
            y = n,
            fill = Surveyed,
            group = Surveyed)) +
          geom_bar(stat="identity", position="dodge") +
          theme_minimal() +
          ylab("Density") +
          xlab(i)
      } else {
        p <- ggplot(
          data.frame(
            Surveyed = srv,
            Variable = z),
          aes(
            x = Variable,
            fill = Surveyed,
            color = Surveyed)) +
          geom_density(alpha = 0.5) +
          geom_rug(data = data.frame(
            Surveyed = srv,
            Variable = z)[!srv,], sides="b") +
          geom_rug(data = data.frame(
            Surveyed = srv,
            Variable = z)[srv,], sides="t") +
          theme_minimal() +
          ylab("Density") +
          xlab(i)

      }
    }
    if (type == "fit") {
      if (CAT) {
        p <- ggplot(
          dat,
          aes(
            y = y,
            x = z)) +
          geom_boxplot(fill = "grey") +
          geom_line(aes(x = as.integer(z), y = zhat),
                      data = aggregate(dat[,c("y", "zhat")],
                                       list(z=dat$z), mean),
                      col = 4) +
          theme_minimal() +
          ylab("Total Moose") +
          xlab(i)
      } else {
        p <- ggplot(
          dat,
          aes(
            y = y,
            x = z)) +
          geom_point() +
          geom_line(aes(x = z, y = zhat), col = 4) +
          theme_minimal() +
          ylab("Total Moose") +
          xlab(i)
      }
    }
    if (type == "map") {
      if (CAT) {
        p <- ggplot(data = x,
          aes_string(x = "CENTRLON", y = "CENTRLAT", fill = i)) +
          geom_tile() +
          coord_map() +
          scale_fill_discrete() +
          theme_minimal()
      } else {
        p <- ggplot(data = x,
          aes_string(x = "CENTRLON", y = "CENTRLAT", fill = i)) +
          geom_tile() +
          coord_map() +
          scale_fill_binned(type = "viridis") +
          theme_minimal()
      }
    }
    p
}

x <- read.csv("inst/extdata/MayoMMU_QuerriedData.csv")
switch_response("total")
x <- mc_update_total(x)

## continuous
i <- "SubShrub250_Fire8212DEM815"
mc_plot_univariate(i, x, "ZINB")
.plot_univariate(i, x, dist="ZINB", type = "density")
.plot_univariate(i, x, dist="ZINB", type = "map")
.plot_univariate(i, x, dist="ZINB", type = "fit")

## discrete
i <- "ZZZ"
x$ZZZ <- cut(x$SubShrub250_Fire8212DEM815, 3)
mc_plot_univariate(i, x, "ZINB")
.plot_univariate(i, x, dist="ZINB", type = "density")
.plot_univariate(i, x, dist="ZINB", type = "map")
.plot_univariate(i, x, dist="ZINB", type = "fit")
```

psolymos commented:
The package will fails checks due to NSE and we'll have to import `rlang::.data` and use inside the function, I'll work on it later.

psolymos commented:
Here is the other function that is to be used to explore fit based on the PI object:

```R
mc_plot_predfit <- function(i, PI) {
  dat <- PI$data[,c("Cell.mean", "Cell.PIL", "Cell.PIU")]
  dat$Surveyed <- PI$data$srv
  dat$z <- PI$data[[i]]
  p <- ggplot(dat, aes(x=z, y=Cell.mean, col=Surveyed)) +
    geom_point() +
    theme_minimal() +
    xlab(i) +
    ylab("TotalMoose")
  p
}
i <- "SubShrub250_Fire8212DEM815"
mc_plot_predfit(i, PI)
```

steffilazerte commented:
I'll add those functions, no problem, also I'm very familiar with fixing NSE checks, so I'm happy to adjust the code as I go, especially as I'll be making my own changes for ggiraph.

steffilazerte committed:
Add interactivity to univariate exploration

- Add function for ggplot
- Add interactive option for ggiraph
- Add workflow examples
- Update UI accordingly
- Add dependencies

steffilazerte commented on R/explore.R:
```diff
@@ -121,6 +121,220 @@ mc_plot_univariate <- function(i, x, dist="ZINB") {
+      tot$z_int <- as.integer(tot$z)
+
+      if(interactive) {
+        out <- dplyr::group_by(dat, .data$z)
+        out <- dplyr::mutate(out,
+                             IQR = stats::IQR(.data$y),
+                             q75 = stats::quantile(.data$y, 0.75),
+                             q25 = stats::quantile(.data$y, 0.25))
+        out <- dplyr::filter(out,
+                             .data$y > (.data$q75 + (1.5 * .data$IQR)) |
+                               .data$y < (.data$q25 - (1.5 * .data$IQR)))
+        out <- dplyr::mutate(out, tt = paste0("SU_ID: ", .data$SU_ID,
+                                              "<br>", .env$i, ": ", .data$z,
+                                              "<br>Total Moose: ", .data$y))
+        dat <- dplyr::mutate(dat, tt = .data$z)
+        tot <- dplyr::mutate(tot, tt = paste0(.data$z,
+                                              "<br>Mean: ", round(.data$zhat, 2)))
+
+        p <- p +
+          ggiraph::geom_boxplot_interactive(
+            data = dat, fill = "grey", outlier.shape = NA, # omit outliers
+            ggplot2::aes(tooltip = .data$tt, data_id = .data$tt)) +
+          ggiraph::geom_point_interactive(                 # manually add outliers
+            data = out, size = 3, alpha = 0.75,
+            ggplot2::aes(tooltip = .data$tt, data_id = .data$tt)) +
+          ggiraph::geom_point_interactive(
+            data = tot, size = 3, col = 4,
+            ggplot2::aes(tooltip = .data$tt, data_id = .data$tt))
```
To make the boxplot outlier point tooltips useful, I had to calculate outliers by hand and plot them separately with geom_point_interactive().

psolymos replied:
I love these plots. So fancy!

steffilazerte commented on R/explore.R:
```diff
@@ -121,6 +121,220 @@ mc_plot_univariate <- function(i, x, dist="ZINB") {
+                                      fill = .data[[i]])) +
+      ggplot2::coord_map() +
+      ggplot2::theme_minimal() +
+      ggplot2::theme(legend.position = "top",
+                     legend.title = ggplot2::element_text(size = 9),
+                     legend.box.margin = ggplot2::margin(0, 0, -20, 0),
+                     plot.margin = ggplot2::unit(c(0,0,0,0), "pt"))
```
To trim white space, note the negative margin to get the plot legend a bit closer to the top of the map.

steffilazerte commented on R/explore.R:
```diff
@@ -121,6 +121,220 @@ mc_plot_univariate <- function(i, x, dist="ZINB") {
+  if (type == "map") {
+    p <- ggplot2::ggplot(data = x,
+                         ggplot2::aes(x = .data$CENTRLON, y = .data$CENTRLAT,
+                                      fill = .data[[i]])) +
```
Use the `.data[[i]]` notation instead of `aes_string` (I think that's less recommended in recent years, but not sure if it's going anywhere).

steffilazerte commented:
The univariate interactive plots are ready. I took some license with the first plot's colours, point sizes and alpha to improve visibility and access to the tooltips. Let me know what you think! 

I haven't been updating the tests because I feel like you're not using them? There is another way of doing tests without `shinytest` or snapshots which make them much smaller and faster, which is a bit more restricted, but can still be useful (something I learned recently: https://mastering-shiny.org/scaling-testing.html#testing-reactivity).

psolymos commented:
I don't have strong feelings about the reactivity testing at this point. Let me thing about it.

Here are 2 tasks:

(1) On the Composition > Summary tab the subsetting works fine. I was wondering if we could add a summary above or below the cell level prediction summary. The same UI element as we have on the Composition > Prediction Interval tab under Summary, the only difference is that it would be for the subset only. I.e. `pred_density_moose_CPI(subset_CPI_data(CPI, 1:100))`.

(2) Regarding the fit:

> Would it be possible to produce this graph as another way to explore predictions. I usually do this in Excel because it's a nice way to see how our model predictions compare to what we counted. I always use the Final RSF covariate (but this could be selected from a drop-down). Total moose counted or predicted on the Y axis and RSF on the X axis.  PI bars would also be nice. 

We can use the `mc_plot_predfit()` function I added above in a comment. I can see this to be added to the Total > Explore Predictions under the table:

- dropdown for covariates (argument `i`)
- and an interactive ggiraph plot version

After these 2 I would merge and send it for testing :)

steffilazerte committed:
Add Total PI exploratory plot

steffilazerte committed:
Add note about model results to total pi

steffilazerte committed:
Update docs properly (i.e. Rmd then knit to md). Add docs for new plot

steffilazerte committed:
Add note about models to composition tab

steffilazerte commented:
Hi @psolymos,

I've got a couple more changes for you: 

1. Predictions plot is now on the Explore Predictions pane
    - Users can only select explanatory variables that are in the models used.
    - You asked for me to put it under the data table, but that seemed pretty far down, so I put it under a "Plot" tab, let me know if you'd prefer I move it down to where you asked :)
    - In the notes, they said having the PI intervals would be great and it looked like you were about to add them to the figure but didn't. I added them (only for the predicted values). Was there a reason I shouldn't have?
    - In general, what do you think?

2. While working on the above, I realized that there's no way of knowing what models the PI results are being displayed for (because you can change the inputs and not update the results).
    - I added a note to each PI tab that states which model(s) are displayed and whether or not they were averaged.
    - However, shouldn't the models used, whether they were averaged or not, and their explanatory variables (and whether they were modelled as count or zero-inflated) be output in the results, for reproducibility?

I'll work next on adding the subset summary. 

Don't we still need to split the PI diagnostics figures so they can be individually copy/pasted?

steffilazerte committed:
Add subset summary to Comp PI

steffilazerte commented:
Okay, the subset summary went pretty fast, there's now a subset composition summary under the datatable on the Composition > Summary pane.

Enjoy!

psolymos commented:
@steffilazerte thanks for the additions, all look pretty neat. Here are my comments:

> 1. Predictions plot is now on the Explore Predictions pane:
>   - Users can only select explanatory variables that are in the models used.

Sometimes it can be useful to select a variable that is not part of the models, for example expert-based suitability, that we might or might not use in the model, but it can be a part of checking and validation. This restriction can be circumvented by adding a "saturated" model so that all variables become available, and the RSF that was mentioned is often part of the models, so this setup is OK 

> - You asked for me to put it under the data table, but that seemed pretty far down, so I put it under a "Plot" tab, let me know if you'd prefer I move it down to where you asked :)

Sounds good to me :)

> - In the notes, they said having the PI intervals would be great and it looked like you were about to add them to the figure but didn't. I added them (only for the predicted values). Was there a reason I shouldn't have?

I like the way you implemented the PI. I left it out because I was imagining a vertical line through the dots (not as a ribbon in the background) which seemed too busy. Thanks!

> - In general, what do you think?

Amazing work!

> 2. While working on the above, I realized that there's no way of knowing what models the PI results are being displayed for (because you can change the inputs and not update the results).
> - I added a note to each PI tab that states which model(s) are displayed and whether or not they were averaged.

Great. I agree, users might loose sight of the actual model being displayed in PI.

> - However, shouldn't the models used, whether they were averaged or not, and their explanatory variables (and whether they were modelled as count or zero-inflated) be output in the results, for reproducibility?

Yes good point. I'll return the list including the inputs so that it can be tracked. Having the model saved could be important.
